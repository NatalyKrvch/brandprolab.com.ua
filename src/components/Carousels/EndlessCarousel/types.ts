export interface EndlessCarouselProps {
  speed?: number;
  pauseOnHover?: boolean;
  gradient?: boolean;
  className?: string;
}
