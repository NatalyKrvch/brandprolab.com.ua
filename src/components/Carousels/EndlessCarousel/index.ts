export { default as EndlessCarousel } from './EndlessCarousel';
export * from './types';
