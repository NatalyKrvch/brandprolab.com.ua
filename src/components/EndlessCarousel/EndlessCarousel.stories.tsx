import type { Meta, StoryObj } from '@storybook/react';

import EndlessCarousel from './EndlessCarousel';

const meta: Meta<typeof EndlessCarousel> = {
  title: 'Components/EndlessCarousel',
  component: EndlessCarousel,
  parameters: {
    layout: 'fullscreen',
  },
};

export default meta;

type Story = StoryObj<typeof EndlessCarousel>;

export const Default: Story = {
  args: {
    children: Array.from({ length: 5 }, (_, i) => (
      <div
        key={i}
        className="bg-primary mx-4 rounded-md px-6 py-4 text-white shadow"
        data-testid="carousel-item"
      >
        Відгук #{i + 1}
      </div>
    )),
  },
};

export const FastAndGradient: Story = {
  args: {
    speed: 100,
    gradient: true,
    children: Array.from({ length: 5 }, (_, i) => (
      <div
        key={i}
        className="bg-secondary mx-4 rounded-md px-6 py-4 text-white shadow"
      >
        Відгук #{i + 1}
      </div>
    )),
  },
};
