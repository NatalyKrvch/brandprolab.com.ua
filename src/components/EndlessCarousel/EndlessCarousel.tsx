import { PropsWithChildren } from 'react';
import Marquee from 'react-fast-marquee';

import {
  ENDLESS_CAROUSEL_CONTENT_TEST_ID,
  ENDLESS_CAROUSEL_TEST_ID,
} from '@/lib/testIDs';

import { EndlessCarouselProps } from './types';

const EndlessCarousel = ({
  children,
  speed = 50,
  pauseOnHover = true,
  gradient = false,
  className = '',
}: PropsWithChildren<EndlessCarouselProps>) => {
  return (
    <div
      data-testid={ENDLESS_CAROUSEL_TEST_ID}
      className={`overflow-hidden ${className}`}
    >
      <Marquee
        data-testid={ENDLESS_CAROUSEL_CONTENT_TEST_ID}
        gradient={gradient}
        speed={speed}
        pauseOnHover={pauseOnHover}
      >
        {children}
      </Marquee>
    </div>
  );
};

export default EndlessCarousel;
