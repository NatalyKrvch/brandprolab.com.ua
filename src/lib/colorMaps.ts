import {
  MainSocialCardColor,
  socialMediaColorClassMapType,
} from '@/components/Cards/SocialMediaCard/types';

export const socialMediaColorClassMap: socialMediaColorClassMapType = {
  [MainSocialCardColor.RED]: {
    bg: 'bg-red-default',
    text: 'text-red-default',
  },
  [MainSocialCardColor.BLUE]: {
    bg: 'bg-blue-default',
    text: 'text-blue-default',
  },
};
