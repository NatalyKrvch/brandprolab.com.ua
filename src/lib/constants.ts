export const ALL_DATA_QUERY = `
{
  "header": *[_type == "header"][0],
  "hero": *[_type == "hero"][0],
  "cases": *[_type == "cases"][0],
  "services": *[_type == "services"][0],
  "testimonials": *[_type == "testimonialsSection"][0],
  "subscribe": *[_type == "subscribeSection"][0],
  "video": *[_type == "videoSection"][0],
  "expertise": *[_type == "expertiseSection"][0],
  "diagnostic": *[_type == "diagnosticSection"][0],
  "footer": *[_type == "footerSection"][0]
}
`;

export const YOUTUBE_REGEX =
  /(?:youtu\.be\/|youtube\.com\/(?:.*[?&]v=|embed\/|v\/|shorts\/|live\/))([^&?/]+)/;

export enum IconClassType {
  SERVICES = 'services',
  EXPERTISE = 'expertise',
  SOCIAL_MEDIA = 'social-media',
  SOCIAL_MEDIA_MAIN = 'social-media-main',
}
export const REVALIDATE = 60;
export const ENDLESS_CAROUSEL_SPEED = 50;
export const SLIDER_RIGHT_SHIFT = 2;
export const SLIDER_LEFT_SHIFT = -1;
export const AMOUNT_OF_TESTIMONIALS_WORDS = 15;
export const AMOUNT_OF_CASES_WORDS_DESKTOP = 13;
export const AMOUNT_OF_CASES_WORDS_TABLET = 9;
export const AMOUNT_OF_CASES_WORDS_MOBILE = 13;
