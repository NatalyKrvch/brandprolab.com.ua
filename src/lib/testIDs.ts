export const INTERNAL_LINK_TEST_ID = 'internal-link-test-id';
export const EXTERNAL_LINK_TEST_ID = 'external-link-test-id';
export const FULL_WIDTH_CONTAINER_TEST_ID = 'full-width-container-test-id';
export const LIMITED_WIDTH_CONTAINER_TEST_ID =
  'limited-width-container-test-id';
export const LEFT_LIMITED_CONTAINER_TEST_ID = 'left-limited-container-test-id';
export const ICON_COMPONENT_TEST_ID = 'icon-component-test-id';
export const ICON_COMPONENT_IMAGE_TEST_ID = 'icon-component-image-test-id';
export const LABEL_COMPONENT_TEST_ID = 'label-component-test-id';
export const BUTTON_TEST_ID = 'button-test-id';
export const UNDERLINED_TEXT_TEST_ID = 'underlined-text-test-id';
export const USER_BADGE_TEST_ID = 'user-badge-test-id';
export const USER_BADGE_IMAGE_TEST_ID = 'user-badge-image-test-id';
export const USER_BADGE_NAME_TEST_ID = 'user-badge-name-test-id';
export const TESTIMONIAL_CARD_TEST_ID = 'testimonials-card-test-id';
export const TESTIMONIAL_TEXT_TEST_ID = 'testimonials-text-test-id';
export const LIST_TEST_ID = 'list-test-id';
export const LIST_ITEM_TEST_ID = 'list-item-test-id';
export const SIMPLE_CARD_TEST_ID = 'simple-card-test-id';
export const DIPLOMA_CARD_TEST_ID = 'diploma-card-test-id';
export const DIPLOMA_IMAGE_TEST_ID = 'diploma-image-test-id';
export const DIPLOMA_OVERLAY_TEST_ID = 'diploma-overlay-test-id';
export const PHOTO_CARD_IMAGE_TEST_ID = 'photo-card-image-test-id';
export const PHOTO_CARD_TEXT_TEST_ID = 'photo-card-text-test-id';
export const VIDEO_IFRAME_TEST_ID = 'video-iframe-test-id';
export const VIDEO_PLAY_BUTTON_TEST_ID = 'video-play-button-test-id';
export const VIDEO_TEST_ID = 'video-test-id';
export const VIDEO_THUMBNAIL_TEST_ID = 'video-thumbnail-test-id';
export const FLIP_CARD_BACK_TEST_ID = 'flip-card-back-test-id';
export const FLIP_CARD_FRONT_TEST_ID = 'flip-card-front-test-id';
export const FLIP_CARD_TEST_ID = 'flip-card-test-id';
export const SERVICE_CARD_FRONT_DESCRIPTION_TEST_ID =
  'service-card-front-description-test-id';
export const SERVICE_CARD_FRONT_HEADER_TEST_ID =
  'service-card-front-header-test-id';
export const SERVICE_CARD_BACK_HEADER_TEST_ID =
  'service-card-back-header-test-id';
export const SERVICE_CARD_BACK_LIST_ITEM_TEST_ID =
  'service-card-back-list-item-test-id';
export const SERVICE_CARD_BACK_LIST_TEST_ID = 'service-card-back-list-test-id';
export const SOCIAL_MEDIA_LABEL_TEST_ID = 'social-media-label-test-id';
export const SOCIAL_MEDIA_DESCRIPTION_TEST_ID =
  'social-media-description-test-id';
export const FOOTER_SECTION_TEST_ID = 'footer-section-test-id';
export const PAGINATION_DOT_TEST_ID = 'pagination-dot-test-id';
export const CAROUSEL_NAV_WRAPPER_TEST_ID = 'carousel-nav-wrapper-test-id';
export const CAROUSEL_NAV_NEXT_TEST_ID = 'carousel-nav-next-test-id';
export const CAROUSEL_NAV_PREV_TEST_ID = 'carousel-nav-prev-test-id';
export const ENDLESS_CAROUSEL_TEST_ID = 'endless-carousel-test-id';
export const ENDLESS_CAROUSEL_CONTENT_TEST_ID =
  'endless-carousel-content-test-id';
export const MAIN_SPRAY_BG_TEST_ID = 'main-spray-bg-test-id';
export const HERO_MAIN_CARD_ICON_TEST_ID = 'hero-main-card-icon-test-id';
export const HERO_TEXT_CONTENT_TEST_ID = 'hero-text-content-test-id';
export const HERO_TEXT_CONTENT_ICON_TEST_ID = 'hero-text-content-icon-test-id';
export const EXPERTISE_SECTION_TEST_ID = 'expertise-section-test-id';
export const SCROLL_UP_BUTTON_TEST_ID = 'scroll-up-button-test-id';
export const MODAL_WRAPPER_TEST_ID = 'modal-wrapper-test-id';
export const MODAL_CONTENT_TEST_ID = 'modal-content-test-id';
export const MODAL_CLOSE_BUTTON_TEST_ID = 'modal-close-button-test-id';
export const CASE_MODAL_TEXT_TEST_ID = 'case-modal-text-test-id';
export const TESTIMONIAL_MODAL_NAME_TEST_ID = 'testimonial-modal-name-test-id';
export const TESTIMONIAL_MODAL_POSITION_TEST_ID =
  'testimonial-modal-position-test-id';
export const TESTIMONIAL_MODAL_TEXT_TEST_ID = 'testimonial-modal-text-test-id';
export const TESTIMONIAL_MODAL_IMAGE_TEST_ID =
  'testimonial-modal-image-test-id';
