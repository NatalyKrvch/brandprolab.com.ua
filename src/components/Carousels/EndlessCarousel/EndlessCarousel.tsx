import type { PropsWithChildren } from 'react';
import Marquee from 'react-fast-marquee';

import {
  ENDLESS_CAROUSEL_CONTENT_TEST_ID,
  ENDLESS_CAROUSEL_TEST_ID,
} from '@/lib/testIDs';
import { ENDLESS_CAROUSEL_SPEED } from '@/styles/constants';

import type { EndlessCarouselProps } from './types';

const EndlessCarousel = ({
  children,
  speed = ENDLESS_CAROUSEL_SPEED,
  pauseOnHover = true,
  gradient = false,
  className = '',
}: PropsWithChildren<EndlessCarouselProps>) => {
  return (
    <div
      data-testid={ENDLESS_CAROUSEL_TEST_ID}
      className={`overflow-hidden ${className}`}
      aria-roledescription="carousel"
      role="group"
    >
      <Marquee
        data-testid={ENDLESS_CAROUSEL_CONTENT_TEST_ID}
        gradient={gradient}
        speed={speed}
        pauseOnHover={pauseOnHover}
        play={true}
      >
        {children}
      </Marquee>
    </div>
  );
};

export default EndlessCarousel;
