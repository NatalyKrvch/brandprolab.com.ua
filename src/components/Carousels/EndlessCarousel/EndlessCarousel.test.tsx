import { render, screen, waitFor } from '@testing-library/react';

import { ENDLESS_CAROUSEL_TEST_ID } from '@/lib/testIDs';

import EndlessCarousel from './EndlessCarousel';

beforeAll(() => {
  global.ResizeObserver = class {
    observe() {}
    unobserve() {}
    disconnect() {}
  };
});

describe('EndlessCarousel', () => {
  it('renders children inside the marquee', async () => {
    const { container } = render(
      <EndlessCarousel>
        <div data-testid="child">Test Content</div>
      </EndlessCarousel>,
    );

    await waitFor(() => {
      const children = screen.getAllByTestId('child');
      expect(children.length).toBeGreaterThanOrEqual(2);
    });

    expect(container).toMatchSnapshot();
  });

  it('applies custom className to wrapper', () => {
    const { container } = render(
      <EndlessCarousel className="custom-class">
        <div>Content</div>
      </EndlessCarousel>,
    );

    const wrapper = screen.getByTestId(ENDLESS_CAROUSEL_TEST_ID);
    expect(wrapper).toHaveClass('custom-class');
    expect(container).toMatchSnapshot();
  });

  it('respects default props if not provided', () => {
    const { container } = render(
      <EndlessCarousel>
        <div>Content</div>
      </EndlessCarousel>,
    );

    expect(container.querySelector('.rfm-marquee')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
