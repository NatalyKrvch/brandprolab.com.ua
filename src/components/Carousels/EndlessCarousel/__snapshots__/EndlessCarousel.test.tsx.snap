// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EndlessCarousel applies custom className to wrapper 1`] = `
<div>
  <div
    aria-roledescription="carousel"
    class="overflow-hidden custom-class"
    data-testid="endless-carousel-test-id"
    role="group"
  >
    <div
      class="rfm-marquee-container "
      style="--pause-on-hover: paused; --pause-on-click: paused; --width: 100%; --transform: none;"
    >
      <div
        class="rfm-marquee"
        style="--play: running; --direction: normal; --duration: 0s; --delay: 0s; --iteration-count: infinite; --min-width: 100%;"
      >
        <div
          class="rfm-initial-child-container"
        >
          <div
            class="rfm-child"
            style="--transform: none;"
          >
            <div>
              Content
            </div>
          </div>
        </div>
      </div>
      <div
        class="rfm-marquee"
        style="--play: running; --direction: normal; --duration: 0s; --delay: 0s; --iteration-count: infinite; --min-width: 100%;"
      >
        <div
          class="rfm-child"
          style="--transform: none;"
        >
          <div>
            Content
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`EndlessCarousel renders children inside the marquee 1`] = `
<div>
  <div
    aria-roledescription="carousel"
    class="overflow-hidden "
    data-testid="endless-carousel-test-id"
    role="group"
  >
    <div
      class="rfm-marquee-container "
      style="--pause-on-hover: paused; --pause-on-click: paused; --width: 100%; --transform: none;"
    >
      <div
        class="rfm-marquee"
        style="--play: running; --direction: normal; --duration: 0s; --delay: 0s; --iteration-count: infinite; --min-width: 100%;"
      >
        <div
          class="rfm-initial-child-container"
        >
          <div
            class="rfm-child"
            style="--transform: none;"
          >
            <div
              data-testid="child"
            >
              Test Content
            </div>
          </div>
        </div>
      </div>
      <div
        class="rfm-marquee"
        style="--play: running; --direction: normal; --duration: 0s; --delay: 0s; --iteration-count: infinite; --min-width: 100%;"
      >
        <div
          class="rfm-child"
          style="--transform: none;"
        >
          <div
            data-testid="child"
          >
            Test Content
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`EndlessCarousel respects default props if not provided 1`] = `
<div>
  <div
    aria-roledescription="carousel"
    class="overflow-hidden "
    data-testid="endless-carousel-test-id"
    role="group"
  >
    <div
      class="rfm-marquee-container "
      style="--pause-on-hover: paused; --pause-on-click: paused; --width: 100%; --transform: none;"
    >
      <div
        class="rfm-marquee"
        style="--play: running; --direction: normal; --duration: 0s; --delay: 0s; --iteration-count: infinite; --min-width: 100%;"
      >
        <div
          class="rfm-initial-child-container"
        >
          <div
            class="rfm-child"
            style="--transform: none;"
          >
            <div>
              Content
            </div>
          </div>
        </div>
      </div>
      <div
        class="rfm-marquee"
        style="--play: running; --direction: normal; --duration: 0s; --delay: 0s; --iteration-count: infinite; --min-width: 100%;"
      >
        <div
          class="rfm-child"
          style="--transform: none;"
        >
          <div>
            Content
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
